### ISSUE TITLE
Vsync broken in GlfwWindow

### ISSUE THREAD

**Issue by leecannon** [Ultz/Silk.NET#93]
# Summary
Vsync is the wrong way round.

When creating a window and setting vsync to on the framerate is unlimited and setting vsync off limits the framerate to the refresh rate. This is the exact opposite of the desired behaviour.

# Comments
Looking at [GlfwWindow.cs](https://github.com/Ultz/Silk.NET/blob/master/src/Windowing/Silk.NET.Windowing.Desktop/GlfwWindow.cs) in ''RaiseRenderFrame" on line 631 to 641 there is the below code:

```cs
// Identical to RaiseUpdateFrame.
if (FramesPerSecond > double.Epsilon
     && (VSync == VSyncMode.Off || VSync == VSyncMode.Adaptive && IsRunningSlowly))
{
    var sleepTime = _renderPeriod - _renderStopwatch.Elapsed.TotalSeconds;

    if (sleepTime > 0.0)
    {
        Thread.Sleep((int) (1000 * sleepTime));
    }
}
```

Reading the above, unless i'm mistaken if vsync is off or adaptive and running slow, we sleep the thread resulting in the framerate being limited whereas if vsync is on we don't sleep therefore the framerate is unlimited.

Surely this logic is backwards?

**Comment by Perksey** [comment]
Pinging @devvoid, the original author of this code. As far as I'm aware when VSync is enabled GLFW should swap the frame automatically, but it seems that this isn't the case as Silk.NET will still swap the frame anyway.

Will wait for a response from Void.

**Comment by devvoid** [comment]
The block you posted should only be running if you're using a capped framerate without vsync. What is UpdatesPerSecond/FramesPerSecond in your example?

As for vsync not limiting the framerate, that's definitely a bug on our end. Looks like someone put a check to see if our window is running first on the setter for the vsync mode, but the initial vsync is set before that's set to true. Will have a patch up and ready for this soon.

**Comment by leecannon** [comment]
In [this](https://gist.github.com/leecannon/ea4ad121a274426e1761aaa1f4ebb23d) gist i've created a small demo program that checks the FPS for various configurations (just change the `RunType` const):

- Vsync On No FPS
- Vsync On 60 FPS
- Vsync Off No FPS
- Vsync On 60 FPS
- Adaptive No FPS
- Adaptive 60 FPS

The only configuration out of all of these that runs at 60fps is "Vsync Off 60 FPS" all of the others run at thousands of FPS. My displays are both at 60hz.

At first I thought my FPS calculation was to blame but as one of these configurations does hit 60 spot on I don't think that is it.

I have also tried the code in the gist with the 'vsync_fix' branch, the results are similar except rather that running at 60fps "Vsync Off 60 FPS" runs at 30

**Comment by Perksey** [comment]
What happens if you disable automatic buffer swapping in the window options?

**Comment by devvoid** [comment]
I've tried a bunch of different prospective fixes, but nothing worked at all. I'm not sure what's going on, or how to fix it.

If anyone else has any ideas, I would advise you to give them a try, because I can't figure anything out.

**Comment by Pyrdacor** [comment]
I will have a look and do some testing.

I saw that there is a sleep in update and render logic as well. Won't that lead to double delay for single-threaded silk windows?

**Comment by Pyrdacor** [comment]
@leecannon In my current fix #109 I tested your code and now it seem to work correctly. I only have to test adaptive vsync. At least it works with high frame rates (in which case vsync is on).
